File: LeetCode/0155. Min Stack.cpp
/*
0155. Min Stack

实现一个最小栈，要在常数时间内获取最小值。

主要点在于，栈每次弹出时，如何更新最小值。
注意：在每次更新最小值时，在更新函数的开头，原来错误地将minVal设为stackList[0]，但如果遇到栈pop空的情况，minVal就会遗留为stackList[0]。

*/

class MinStack {
public:
    /** initialize your data structure here. */
    int minVal;
    int stackList[10000];
    int headPos;

    MinStack() {
        minVal = 2147483647;
        headPos = 0;
    }

    void push(int x) {
        stackList[headPos++] = x;
        if (x < minVal)
            minVal = x;
    }

    void pop() {
        headPos = headPos - 1;
        updateMin();
    }

    int top() {
        return stackList[headPos - 1];
    }

    int getMin() {
        return minVal;
    }

    void updateMin() {
        // 这里原来错误地将minVal设为stackList[0]，但如果遇到栈pop空的情况，minVal就会遗留为stackList[0]。
        minVal = -2147483647;
        for (int i = 0; i < headPos; i++) {
            if (stackList[i] <= minVal) {
                minVal = stackList[i];
            }
        }
    }
};

File: LeetCode/0155. Min Stack.py
'''
0155. Min Stack

思路：和第一版类似，关键是由于题目要求在常数时间找到最小元素，因此就在pop时花时间维护存储最小元素的变量。

'''

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.s = list()
        # 最小值，最小值的数有几个
        self.mini_count = [float('inf'), 0]


    def findMin(self):
        new_min = min(self.s)
        new_count = self.s.count(new_min)
        return [new_min, new_count]
        

    def push(self, x: int) -> None:
        self.s.append(x)
        if x < self.mini_count[0]:
            self.mini_count[0] = x
            self.mini_count[1] = 1
        elif x == self.mini_count[0]:
            self.mini_count[1] += 1


    def pop(self) -> None:
        delete = self.s.pop()
        if delete == self.mini_count[0]:
            if self.mini_count[1] == 1:
                if len(self.s) == 0:
                    self.mini_count = [float('inf'), 0]
                else:
                    self.mini_count = self.findMin()
            else:
                self.mini_count[1] -= 1


    def top(self) -> int:
        return self.s[-1]

    def getMin(self) -> int:
        return self.mini_count[0]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()